lib/Conversion/HWToBTOR2/HWToBTOR2.cpp

```diff
@@ -0,0 +1,809 @@
+
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    // Build and return the state instruction
+    os << lid++ << wsStr << stateStr << wsStr << sid << wsStr << name << nlStr;
+  }
+
+  // Generates a next instruction, given a width, a state LID, and a next value
+  // LID
+  void genNext(Value next, Operation *reg, int64_t width) {
+    // Retrieve the lid associated with the sort (sid)
+    size_t sid = sortToLIDMap.at(width);
+
+    // Retrieve the LIDs associated to reg and next
+    size_t regLID = getOrCreateOpLID(reg);
+    size_t nextLID = getOrCreateOpLID(next);
+
+    // Build and return the next instruction
+    os << lid++ << wsStr << nextStr << wsStr << sid << wsStr << regLID << wsStr
+       << nextLID << nlStr;
+  }
+
+  // Verifies that the sort required for the given operation's btor2 emission
+  // has been generated
+  int64_t requireSort(mlir::Type type) {
+    // Start by figuring out what sort needs to be generated
+    int64_t width = hw::getBitWidth(type);
+    assert(width != noWidth);
+
+    // Generate the sort regardles of resulting width (nothing will be added if
+    // the sort already exists)
+    genSort(bitvecStr, width);
+    return width;
+  }
+
+  // Generates the transitions requires to finilize the register to state
+  // transition system conversion
+  void finilizeRegVisit(Operation *op) {
+    // Check the register type (done to support non-firrtl registers as well)
+    auto reg = dyn_cast<seq::FirRegOp>(op);
+    if (!reg)
+      return;
+
+    // Generate the reset condition (for sync & async resets)
+    // We assume for now that the reset value is always 0
+    size_t width = hw::getBitWidth(reg.getType());
+    genSort(bitvecStr, width);
+    genZero(width);
+
+    // Extract the `next` operation for each register (used to define the
+    // transition). We need to check if next is a port to avoid nullptrs
+    Value next = reg.getNext();
+
+    // Next should already be associated to an LID at this point
+    // As we are going to override it, we need to keep track of the original
+    // instruction
+    size_t nextLID = noLID;
+
+    // Check for special case where next is actually a port
+    // To do so, we start by checking if our operation is a block argument
+    if (BlockArgument barg = dyn_cast<BlockArgument>(next)) {
+      // Extract the block argument index and use that to get the line number
+      size_t argIdx = barg.getArgNumber();
+
+      // Check that the extracted argument is in range before using it
+      if (inputLIDs.contains(argIdx)) {
+        nextLID = inputLIDs[argIdx];
+      }
+    } else {
+      nextLID = getOrCreateOpLID(next);
+    }
+
+    // Sanity check
+    assert(nextLID != noLID);
+
+    // Generate the ite for the register update reset condition
+    // i.e. reg <= reset ? 0 : next
+    genIte(next.getDefiningOp(), resetLID, constToLIDMap.at({0, width}),
+           nextLID, width);
```

Comment:
`seq.firreg` has an optional reset operand (you could get `getReset()`) so I think we need to use them. I think it's not always correct to initialize registers into 0. Even for resetless registers I guess btor2 implicitly initializes the states to zeros (Sec 2. "In contrast to Btor, where registers are implicitly zero-initialized and memories are uninitialized").

Reply:
btor and btor2 function a bit differently. States in btor2 are just a representation of the register and don't have a implicit value. The encoding of an implicit reset has to be done when generating the btor2.

Also, forgive my lack of mlir experience,  I'm not quite sure I understand how to extract the numerical value from the getReset() call, any tips on how I could do that?

Reply:
> Also, forgive my lack of mlir experience, I'm not quite sure I understand how to extract the numerical value from the getReset() call, any tips on how I could do that?

@Dobios I think you're probably trying to extract a numerical value for the resetValue, instead of the reset - the reset attribute itself is the 1-bit value that triggers a reset, whereas the resetValue is the value the reg is reset to.

IIRC the resetValue attribute that you provide is an SSA value as opposed to a literal, so extracting a numerical value is tricky. I'm not super familiar with BTOR2, but if you're only able to handle fixed reset values, you could fetch the definition of the reset and check if it's a `hw.constant` op - if so, then just extract the value given there, and I guess otherwise throw an error and clarify that you only support direct `hw.constant`-defined resets. If you are able to handle mutable reset values though (which I guess can just be done manually for synchronous resets by adding a mux on the reg input to the BTOR2 you generate) then you should be able to just lower the logic that defines the reset value to BTOR2 and then provide the output of that logic as the reset value in the BTOR2.

Reply:
@TaoBi22 Thanks for the comment! I think that it made realize a core misunderstanding that had in how resets were handled in hw. I didn't realize that they were considered as defined by arbitrary operations. Now that that makes more sense to me, I was able to fix the implementation to handle arbitrary resets. Thanks!